CA-322008: Stop treating Derive datasources as counters

psafont opened the pull request:
This was creating values unreasonable values in Round-robin Archives.

There is still related behaviour to fix: out-of-range values should be treated as unknowns instead of clamped: they are abnormalities and should be treated as such.

Depends on https://github.com/xapi-project/xcp-rrd/pull/30

psafont committed:
tests: add regression test for CA-322008

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

psafont committed:
CA-322008: do not treat DERIVE sources as COUNTER

Data Sources with derive type are not monotonically increasing like
counter ones and may lead to strange metrics being produced.

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

psafont committed:
CA-322008: Reorganize code for incoming values by type

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

psafont committed:
maintenance: use infix operators for Int64 in rrd

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

psafont committed:
maintenance: whitespace and commented code deletion

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

lindig commented:
Values out of range are now mapped to `nan`. Is that the same as unknown? Is `nan` not suggesting very large or small values? A NaN propagates in expressions to the result - so any derived value is likely to be NaN, too. I really don't know what the expectation here are. The code looks fine in general.

psafont commented:
> Values out of range are now mapped to nan. Is that the same as unknown? Is nan not suggesting very large or small values?

Unkowns in the data-sources are translated to NaN in the Archives. So I think they are equivalent. Yes, nan are supposed to appear when the datapoints are out of range.

> A NaN propagates in expressions to the result - so any derived value is likely to be NaN, too. I really don't know what the expectation here are.

The NaNs in the Data Sources are not allowed to be propagated to the other values: https://github.com/xapi-project/xcp-rrd/pull/31/files#diff-3ce8fe4c2864112c8e27382d2f24ab29R326

lindig commented on lib/rrd.ml:
```diff
@@ -421,7 +427,7 @@ let rrd_create dss rras timestep inittime =
 let rrd_add_ds rrd now newds =
   if List.mem newds.ds_name (ds_names rrd) then rrd else
-    let npdps = Int64.div (Int64.of_float now) rrd.timestep in
+    let npdps = (Int64.of_float now) /// rrd.timestep in
```
`(..)` not required

psafont commented:
This should not be merged, testing found this has a side-effect: xcp-rdd reports all the physical cpus (and the average) to be 100% used (the values are 1).

This happens only with [function](https://github.com/xapi-project/xcp-rrdd/blob/master/rrdd/xcp_rrdd.ml#L236), not with VCPUs nor with C/P-state reporting from xenpm.

psafont commented:
Not sure why get a non-exhaustive match warning:
```
dune runtest
File "lib/rrd.ml", line 245, characters 8-799:
Error (warning 8): this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
Absolute
(However, some guarded clause may match this value.)
```

The PCPU behaviour is now fixed, it was not due to the transform. What's missing now is testing that marking out-of-bounds values as NaN works correctly.

psafont committed:
CA-322008: Report out-of-bounds PDP values as NaN

Previously the values were just clamped before being applied the
transform function from the client.

Now the values marked as unknown values after having the transform
applied, in case the transforms do not check for out-of-bounds values.

Treating out-of-bounds values as Unknown matches rrdtool's behaviour.

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

psafont commented:
Fixed the non-exhaustive pattern-matching due to the `when` guard.

Now the code does not produce unexpected NaNs in the test host.

lindig commented:
This looks carefully done and tested. Can we clean up the commits a bit such that we have consistently the ticket at the front and maybe squash commits?
